pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -95,108 +153,152 @@ public static void init() {
+        Class<?> returnClass = method.getReturnType();
+        RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        RelDataType returnType = returnClass == Object.class ? typeFactory.createSqlType(SqlTypeName.ANY)
+            : JavaTypeFactoryImpl.toSql(typeFactory, typeFactory.createJavaType(returnClass));
+
+        if (!functionInfo.hasNullableParameters()) {
+          // When any parameter is null, return is null
+          for (RelDataType type : opBinding.collectOperandTypes()) {
+            if (type.isNullable()) {
+              return typeFactory.createTypeWithNullability(returnType, true);
+            }
+          }
+        }
+
+        return method.isAnnotationPresent(Nullable.class) ? typeFactory.createTypeWithNullability(returnType, true)
+            : returnType;
+      };
+    }
+
+    private SqlOperandTypeChecker getOperandTypeChecker() {
+      if (_functionInfoMap.containsKey(VAR_ARG_KEY)) {
+        return OperandTypes.VARIADIC;
+      }
+      if (_functionInfoMap.size() == 1) {
+        return getOperandTypeChecker(_functionInfoMap.values().iterator().next().getMethod());
+      }
+      List<SqlOperandTypeChecker> operandTypeCheckers = new ArrayList<>(_functionInfoMap.size());
+      for (FunctionInfo functionInfo : _functionInfoMap.values()) {
+        operandTypeCheckers.add(getOperandTypeChecker(functionInfo.getMethod()));
+      }
+      return OperandTypes.or(operandTypeCheckers.toArray(new SqlOperandTypeChecker[0]));
+    }
+
+    private static SqlTypeFamily getSqlTypeFamily(Class<?> clazz) {
+      // NOTE: Pinot allows some non-standard type conversions such as Timestamp <-> long, boolean <-> int etc. Do not
+      //       restrict the type family for now. We only restrict the type family for String so that cast can be added.
+      //       Explicit cast is required to correctly convert boolean and Timestamp to String.
+      // TODO: Revisit this.
+      return clazz == String.class ? SqlTypeFamily.CHARACTER : SqlTypeFamily.ANY;
+    }
+
+    private static SqlOperandTypeChecker getOperandTypeChecker(Method method) {
+      Class<?>[] parameterTypes = method.getParameterTypes();
+      int length = parameterTypes.length;
+      SqlTypeFamily[] typeFamilies = new SqlTypeFamily[length];
+      for (int i = 0; i < length; i++) {
+        typeFamilies[i] = getSqlTypeFamily(parameterTypes[i]);
+      }
+      return OperandTypes.family(typeFamilies);
     }
 
-    @ScalarFunction(names = {"vectorSimilarity", "vector_similarity"}, isPlaceholder = true)
-    public static boolean vectorSimilarity(float[] vector1, float[] vector2, int topk) {
-      throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
+    @Nullable
+    @Override
+    public FunctionInfo getFunctionInfo(int numArguments) {
+      FunctionInfo functionInfo = _functionInfoMap.get(numArguments);
+      return functionInfo != null ? functionInfo : _functionInfoMap.get(VAR_ARG_KEY);
```

Comment:
Yes. In order to have finer control, it should implement `PinotScalarFunction`, or explicitly register the operator in `PinotOperatorTable`

Reply:
There are 3 types of function implementations:
- AggregationFunction
- TransformFunction
- Annotated ScalarFunction

`FunctionRegistry` contains all the annotated scalar function implementations, and will register the converted function operator into `PinotOperatorTable` so that it is discoverable with multi-stage engine. We allow directly registering function into `PinotOperatorTable` to override the java class based inference. One example is `FROM_DATE_TIME` where we want to return TIMESTAMP in V2; it will return LONG in V1 per the function signature in `DateTimeFunctions.fromDateTime()`. Ideally we should unify them, but that is out of the scope of this PR, thus leaving a TODO.